cmd/pomerium-cli/cli.go

```diff
@@ -0,0 +1,157 @@
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/fatih/color"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+type stringSlice []string
+
+func (i *stringSlice) String() string {
+	return fmt.Sprint(*i)
+}
+
+func (i *stringSlice) Set(value string) error {
+	if len(*i) > 0 {
+		return errors.New("already set")
+	}
+	for _, dt := range strings.Split(value, ",") {
+		*i = append(*i, dt)
+	}
+	return nil
+}
+
+type serviceAccount struct {
+	// Standard claims (as specified in RFC 7519).
+	jwt.Claims
+	// Pomerium claims (not standard claims)
+	Email             string   `json:"email"`
+	Groups            []string `json:"groups,omitempty"`
+	User              string   `json:"user,omitempty"`
+	ImpersonateEmail  string   `json:"impersonate_email,omitempty"`
+	ImpersonateGroups []string `json:"impersonate_groups,omitempty"`
+}
+
+func main() {
+	if err := run(); err != nil {
+		fmt.Fprintf(os.Stderr, color.RedString("\n⛔️ %s\n\n"), err)
+		printHelp(flags)
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+var flags *flag.FlagSet
+
+func run() error {
+	var sa serviceAccount
+
+	// temporary variables we will use to hydrate our service account
+	// struct from basic types pulled in from our flags
+	var (
+		aud               stringSlice
+		groups            stringSlice
+		impersonateGroups stringSlice
+		expiry            time.Duration
+	)
+
+	// set our JWT claims from the supplied CLI flags
+	flags = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	flags.StringVar(&sa.Email, "email", "", "Email")
+	flags.StringVar(&sa.ImpersonateEmail, "impersonate_email", "", "Impersonation Email (optional)")
+	flags.StringVar(&sa.Issuer, "iss", "", "Issuing Server (e.g authenticate.int.pomerium.io)")
+	flags.StringVar(&sa.Subject, "sub", "", "Subject (typically User's GUID)")
+	flags.StringVar(&sa.User, "user", "", "User (typically User's GUID)")
+	flags.Var(&aud, "aud", "Audience (e.g. httpbin.int.pomerium.io,prometheus.int.pomerium.io)")
+	flags.Var(&groups, "groups", "Groups (e.g. admins@pomerium.io,users@pomerium.io)")
+	flags.Var(&impersonateGroups, "impersonate_groups", "Impersonation Groups (optional)")
+	flags.DurationVar(&expiry, "expiry", time.Hour, "Expiry")
+
+	// hydrate the sessions non-primate types
+	if err := flags.Parse(os.Args[1:]); err != nil {
+		return err
+	}
+	sa.Audience = jwt.Audience(aud)
+	sa.Groups = []string(groups)
+	sa.ImpersonateGroups = []string(impersonateGroups)
+	sa.Expiry = jwt.NewNumericDate(time.Now().Add(expiry))
+	sa.IssuedAt = jwt.NewNumericDate(time.Now())
+	sa.NotBefore = jwt.NewNumericDate(time.Now())
+
+	if sa.Email == "" {
+		return errors.New("email is required")
+	}
+
+	if len(sa.Audience) == 0 {
+		return errors.New("aud is required")
+	}
+
+	if sa.Issuer == "" {
+		return errors.New("iss is required")
+	}
+
+	// check that we've got our shared key to sign our jwt
+	args := flags.Args()
+	if len(args) != 1 {
+		return errors.New("shared key required")
+	}
+
+	sharedKey := args[0]
```

Comment:
I feel a little odd asking for this key as a command line arg without another option.  Stdin and/or a file path containing the secret would be good alternatives less prone to leaking in shell history, etc.

Reply:
Updated! Still allows insecure-y argument but has option for stdin input.

![carbon](https://user-images.githubusercontent.com/1544881/77497804-754e9000-6e0b-11ea-9982-53161270a92f.png)

Reply:
\cc @freeyoung how's this look?

Reply:
lgtm 👍